vortex-array/src/array/chunked/mod.rs

```diff
@@ -132,6 +133,78 @@ impl ChunkedArray {
     pub fn array_stream(&self) -> impl ArrayStream + '_ {
         ArrayStreamAdapter::new(self.dtype().clone(), stream::iter(self.chunks().map(Ok)))
     }
+
+    pub fn rechunk_default(&self) -> VortexResult<Self> {
+        let sixteen_megabytes = 1 << 24;
+        let chunk_max = 1 << 16; // BtrBlocks uses 64K so we use 64Ki
+        self.rechunk(sixteen_megabytes, chunk_max)
+    }
+
+    pub fn rechunk(&self, max_n_bytes: usize, max_n_elements: usize) -> VortexResult<Self> {
+        fn combine_validities(
+            dtype: &DType,
+            validities: Vec<LogicalValidity>,
+        ) -> VortexResult<Validity> {
+            let validity = validities.clone().into_iter().collect::<Validity>();
+            match (dtype.is_nullable(), validity) {
+                (true, validity) => Ok(validity),
+                (false, Validity::AllValid) => Ok(Validity::NonNullable),
+                (false, _) => vortex_bail!(
+                    "for non-nullable dtype, child validities ought to all be AllValid"
+                ),
+            }
+        }
+
+        let dtype = self.dtype();
+
+        let mut new_chunks: Vec<Array> = Vec::new();
+        let mut validities_to_combine: Vec<LogicalValidity> = Vec::new();
+        let mut chunks_to_combine: Vec<Array> = Vec::new();
+        let mut new_chunk_n_bytes = 0;
+        let mut new_chunk_n_elements = 0;
+        for chunk in self.chunks() {
+            let n_bytes = chunk.nbytes();
+            let n_elements = chunk.len();
+
+            if (new_chunk_n_bytes + n_bytes > max_n_bytes
+                || new_chunk_n_elements + n_elements > max_n_elements)
+                && !chunks_to_combine.is_empty()
+            {
+                let canonical = try_canonicalize_chunks(
```

Comment:
though you can avoid duplicating this logic by constructing chunked array directly